d2/src/main/java/com/linkedin/d2/balancer/dualread/DualReadLoadBalancerMonitor.java

```diff
@@ -19,74 +19,195 @@
+      CacheEntry<T> newEntry = new CacheEntry<>(propertyVersion, getTimestamp(), property);
+      if (!compareEntries(entry, newEntry))
+      {
+        _rateLimitedLogger.warn("Received mismatched properties from dual read. Old LB: {}, New LB: {}",
+            entry, newEntry);
+      }
       cacheToCompare.invalidate(propertyName);
     }
-
-    cacheToAdd.put(propertyName, new CacheEntry<>(propertyVersion, property));
+    else
+    {
+      cacheToAdd.put(propertyName, new CacheEntry<>(propertyVersion, getTimestamp(), property));
+    }
   }
 
+  abstract boolean compareEntries(CacheEntry<T> oldLbEntry, CacheEntry<T> newLbEntry);
+
+  abstract void onEvict();
+
   private Cache<String, CacheEntry<T>> buildCache()
   {
     return CacheBuilder.newBuilder()
         .maximumSize(MAX_CACHE_SIZE)
-        // Invalidate cache entry after 30s to avoid overflow
-        .expireAfterWrite(CACHE_EXPIRATION_SECOND, TimeUnit.SECONDS)
         .removalListener(notification ->
         {
-          // If a cache entry is evicted due to TTL, print a WARN log because it only
+          // If a cache entry is evicted due to maximum capacity, print a WARN log because it only
           // receives update from one source
-          if (notification.getCause().equals(RemovalCause.EXPIRED))
+          if (notification.getCause().equals(RemovalCause.SIZE))
           {
-            _log.warn("Cache entry evicted after {} s: {}", CACHE_EXPIRATION_SECOND, notification.getValue());
+            _rateLimitedLogger.warn("Cache entry evicted since cache is full: {}", notification.getValue());
+            onEvict();
           }
         })
         .build();
   }
 
+  private String getTimestamp() {
+    return ZonedDateTime.ofInstant(Instant.ofEpochMilli(_clock.currentTimeMillis()), ZoneId.systemDefault())
+        .format(_format);
+  }
+
   private static final class CacheEntry<T>
   {
-    final long _version;
+    final String _version;
+    final String _timeStamp;
     final T _data;
 
-    CacheEntry(long version, T data)
+    CacheEntry(String version, String timeStamp, T data)
     {
       _version = version;
+      _timeStamp = timeStamp;
       _data = data;
     }
 
     @Override
     public String toString()
     {
-      return "CacheEntry{" + "_version=" + _version + ", _data=" + _data + '}';
+      return "CacheEntry{" + "_version=" + _version + ", _timeStamp='" + _timeStamp + '\'' + ", _data=" + _data + '}';
+    }
+  }
+
+  public static final class ClusterPropertiesDualReadMonitor extends DualReadLoadBalancerMonitor<ClusterProperties>
+  {
+    private final DualReadLoadBalancerJmx _dualReadLoadBalancerJmx;
+
+    public ClusterPropertiesDualReadMonitor(DualReadLoadBalancerJmx dualReadLoadBalancerJmx, Clock clock)
+    {
+      super(clock);
+      _dualReadLoadBalancerJmx = dualReadLoadBalancerJmx;
+    }
+
+    @Override
+    boolean compareEntries(CacheEntry<ClusterProperties> oldLbEntry, CacheEntry<ClusterProperties> newLbEntry)
```

Comment:
"isEqual" is simpler. Same applies to the Service and Uri comparing functions too.